### PR TITLE
Add support for CorrCholeskyTransform

### PR DESCRIPTION
**Pull request by neerajprad**
This adds a transform to convert a real vector of (D * (D-1))/2 dimension into the cholesky factor of a D x D correlation matrix. This follows the implementation in [NumPyro](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/distributions/transforms.py) by @fehiepsi. This is needed for the LKJDistribution which will be added in a subsequent PR.

Also in line with the ongoing effort to refactor distributions test, this moves the transforms test into its own file that uses pytest with parametrized fixtures. 

For review:
 @fehiepsi - could you help review the math?
 @fritzo - do you have any suggestions for what to do about the event dimension (more details are in the comment below)? 
 @ezyang - could you review the changes in `run_test.py`? Instead of a separate `PYTEST_TESTS`, I have clubbed these tests in `USE_PYTEST_LIST` to avoid duplicate logic. The only difference is that we do not anymore check if pytest is not installed and exclude the tests in the list. I figured that if existing tests are already using pytest, this should not matter. 

TODOs (probably not all can be satisfied at the same time):
 - [x] Use operations that are JIT friendly, i.e. the transform works with different sized input under JIT.
 - [x] Resolve test failures - currently `arange(scalar_tensor)` fails on certain backends but this is needed for JIT. Maybe we should only support same sized tensor under JIT?
 - [x] Add tests to check that the transform gives correct gradients and is in agreement with the `log_det_jacobian`.
 - [x] Add `input_event_dim` and `output_event_dim` to `CorrCholeskyTransform`.

### CODE REVIEW

File: torch/distributions/transforms.py [hunk 1]
```diff
@@ -535,6 +536,83 @@ def log_abs_det_jacobian(self, x, y):
...
+            a. Scales into the interval :math:`(-1, 1)` domain: :math:`r_i = \tanh(X_i)`.
+            b. Transforms into an unsigned domain: :math:`z_i = r_i^2`.
+            c. Applies :math:`s_i = StickBreakingTransform(z_i)`.
+            d. Transforms back into signed domain: :math:`y_i = sign(r_i) * \sqrt{s_i}`.
+    """
+    domain = constraints.real_vector
+    codomain = constraints.corr_cholesky
+    bijective = True
+    # Note that since we add a dim in _call this should actually be 2 for the inverse transform.
+    event_dim = 1
```
Comment (neerajprad) [inline torch/distributions/transforms.py:559]: @fehiepsi, @fritzo - Note that the forward transform add an additional dim, so `event_dim=1` is defined w.r.t. to the value passed to the forward transform. The inverse transform's event_dim is consequently incorrect. I wasn't sure what to do. One thing would be to modify the inverse transform for this to have `event_dim=2`. Any other suggestions?

File: torch/distributions/transforms.py [hunk 2]
```diff
@@ -535,6 +536,83 @@ def log_abs_det_jacobian(self, x, y):
...
+        x = x.clamp(min=-1 + eps, max=1 - eps)
+        n = (1 + math.sqrt(1 + 8 * x.shape[-1])) / 2
+        if round(n) - n > eps:
+            raise ValueError(f'The size of last dimension is {y.shape[-1]} which cannot be expressed ' +
+                             '(D * (D-1)) / 2, i.e. the lower triangular part of a square D x D matrix.')
+        n = round(n)
+        r = torch.zeros(x.shape[:-1] + (n, n))
+        # TODO: Explore a JIT friendly way to convert a tensor to
+        # a lower triangular matrix.
+        tril_idx = torch.tril_indices(n, n, -1, device=x.device)
```
Comment (neerajprad) [inline torch/distributions/transforms.py:573]: Note that `tril_indices` isn't JIT friendly since n cannot be a pytorch scalar, so to make this generalize to arbitrary `n` and work with JIT, we'll need to use something else here.

**Review by fehiepsi — COMMENTED** [review]
The math looks good to me. I just have some minor comments. Besides `event_dim=1`, which needs some discussions, I am a bit worried about the gradient of the transform. Is there any test for it (e.g. test the agreement of log_det implementation and the gradient of `t(x)` where `t` is a transform)? The reason is in an old PyTorch version, gradients might not proprogate properly when taking `z.sqrt()` for a lower triangular matrix `z`. It might have been fixed now but worth to test. :)

File: torch/distributions/transforms.py (hunk 2 above)
Comment (fehiepsi) [reply torch/distributions/transforms.py:573]: I guess we can make two branches: 1 under JIT and one not. In JIT tracing, we can do
```
tril_idx = (r + 1).tril(diagonal=-1) > 0.5  # I'm not sure if this works under JIT...
```

File: torch/distributions/transforms.py [hunk 3]
```diff
@@ -535,6 +536,83 @@ def log_abs_det_jacobian(self, x, y):
...
+    def _call(self, x):
+        x = torch.tanh(x)
+        eps = torch.finfo(x.dtype).eps
+        x = x.clamp(min=-1 + eps, max=1 - eps)
+        n = (1 + math.sqrt(1 + 8 * x.shape[-1])) / 2
+        if round(n) - n > eps:
+            raise ValueError(f'The size of last dimension is {y.shape[-1]} which cannot be expressed ' +
+                             '(D * (D-1)) / 2, i.e. the lower triangular part of a square D x D matrix.')
+        n = round(n)
+        r = torch.zeros(x.shape[:-1] + (n, n))
```
Comment (fehiepsi) [inline torch/distributions/transforms.py:570]: nit: x.new_ones(...)

File: torch/distributions/transforms.py (hunk 1 above)
Comment (fehiepsi) [reply torch/distributions/transforms.py:559]: I am not sure what we can do here... Either leaving this `1` or `2` is fine I guess.

File: torch/distributions/transforms.py [hunk 4]
```diff
@@ -535,6 +536,83 @@ def log_abs_det_jacobian(self, x, y):
...
+        # a lower triangular matrix.
+        tril_idx = torch.tril_indices(n, n, -1, device=x.device)
+        r[..., tril_idx[0], tril_idx[1]] = x
+        # apply stick-breaking on the squared values
+        # Note that y = sign(r) * sqrt(z * z1m_cumprod)
+        #             = (sign(r) * sqrt(z)) * sqrt(z1m_cumprod) = r * sqrt(z1m_cumprod)
+        z = r ** 2
+        z1m_cumprod_sqrt = (1 - z).cumprod(-1).sqrt()
+        # Diagonal elements must be 1.
+        r.diagonal(dim1=-1).copy_(r.new_ones(r.shape[:-2] + (n,)))
```
[the remark this answers is not in the document]
Comment (neerajprad) [reply torch/distributions/transforms.py:581]: Thanks, I didn't know about this function!

[the remark this answers is not in the document]
Comment (neerajprad) [reply torch/distributions/transforms.py:581]: It seems like this won't work for batched tensors.

File: torch/distributions/transforms.py (hunk 2 above)
Comment (neerajprad) [reply torch/distributions/transforms.py:573]: Thanks. Your comment on using masked indexing seems to work with JIT.

File: torch/distributions/transforms.py (hunk 3 above)
Comment (neerajprad) [reply torch/distributions/transforms.py]: Good point, changed to `new_zeros(*)`

**Comment by neerajprad** [comment]
>  I am a bit worried about the gradient of the transform. Is there any test for it (e.g. test the agreement of log_det implementation and the gradient of t(x) where t is a transform)? 

We don't have any gradient tests, but I will add it in this PR.

> The reason is in an old PyTorch version, gradients might not proprogate properly when taking z.sqrt() for a lower triangular matrix z. It might have been fixed now but worth to test. :)

Can you point me to this issue? I think I'm facing this right now.

**Comment by fehiepsi** [comment]
@neerajprad It is [this issue](https://github.com/pytorch/pytorch/issues/15506). Back then, I resolved it with
```python
    z_cumprod_sqrt = z_cumprod.new_zeros(z_cumprod.shape)
    z_cumprod_sqrt[tril_indx] = z_cumprod[tril_indx].sqrt()
```

File: torch/distributions/transforms.py (hunk 1 above)
Comment (fritzo) [reply torch/distributions/transforms.py:559]: Good point, I believe we did not anticipate transforms that have change event dim? How about we add new properties to the base `Transform` class and then override these in your new transform?

<details>

```py
class Transform:
    @property
    def input_event_dim(self):
        return self.event_dim

    @property
    def output_event_dim(self):
        return self.event_dim

class _InverseTransform(Transform):
    @property
    def input_event_dim(self):
        return self._inv.output_event_dim

    @property
    def output_event_dim(self):
        return self._inv.input_event_dim

class CorrCholeskyTransform(Transform):
    input_event_dim = 1
    output_event_dim = 2

    @property
    def event_dim(self):
        raise ValueError("Please use .intput_event_dim or .output_event_dim instead")
```

</details>

Comment (neerajprad) [reply torch/distributions/transforms.py:559]: That sounds like a good idea. I'll make the change!

**Comment by ezyang** [comment]
`run_tests.py` changes look fine. You're testing more now, not less, right? Very simple.

File: torch/distributions/transforms.py (hunk 1 above)
Comment (neerajprad) [reply torch/distributions/transforms.py]: Done. Thanks for the suggestion!

**Comment by neerajprad** [comment]
> We don't have any gradient tests, but I will add it in this PR.

@fehiepsi - Thanks for the push to add tests for jacobian (I caught a bug too). I realized that we had tests for jacobian of univariate transforms but I have extended those to support transforms like CorrCholesky and StickBreaking. I have also abstracted conversion of vector to lower triangular matrix into separate utility functions (as you did in NumPyro) to make it easier to test.

File: test/distributions/test_distributions.py [hunk 5]
```diff
@@ -53,15 +53,10 @@
                                  VonMises, Weibull, constraints, kl_divergence)
 from torch.distributions.constraint_registry import transform_to
 from torch.distributions.constraints import Constraint, is_dependent
+from torch.distributions.transforms import (AffineTransform, CatTransform, ExpTransform,
+                                            StackTransform, identity_transform)
```
Comment (fehiepsi) [inline test/distributions/test_distributions.py:57]: I think this will trigger lint issue: `kl -> transforms`

File: test/distributions/test_transforms.py [hunk 6]
```diff
@@ -0,0 +1,365 @@
+from numbers import Number
+
+import pytest
+
+import torch
+from torch.autograd.functional import jacobian
+from torch.distributions import Dirichlet, Normal, TransformedDistribution, constraints
+from torch.distributions.utils import tril_matrix_to_vec, vec_to_tril_matrix
+from torch.distributions.transforms import (AbsTransform, AffineTransform, ComposeTransform,
```
Comment (fehiepsi) [inline test/distributions/test_transforms.py:9]: lint: `transforms -> utils`

File: torch/distributions/constraints.py [hunk 7]
```diff
@@ -275,6 +276,17 @@ def check(self, value):
...
 
 
+class _CorrCholesky(Constraint):
+    """
+    Constrain to lower-triangular square matrices with positive diagonals and each
+    row vector being of unit length.
+    """
+    def check(self, value):
+        row_norm = torch.norm(value, dim=-1)
+        unit_row_norm = (row_norm <= 1. & row_norm >= 1e-6).all(dim=-1)
```
Comment (fehiepsi) [inline torch/distributions/constraints.py:286]: I guess you meant `1 - 1e-6`.

nit: `torch.norm` is deprecated https://pytorch.org/docs/stable/generated/torch.norm.html

**Review by fehiepsi — COMMENTED** [review]
LGTM overall! I just have a few comments else. You might also want to test the utilities `tril_matrix_to_vec`, `vec_to_tril_matrix` with different `diag` values.

File: test/distributions/test_transforms.py (hunk 6 above)
Comment (neerajprad) [reply test/distributions/test_transforms.py:8]: Interestingly, linter didn't complain. My current IDE setup isn't automatically correcting lint errors, which I will fix shortly. Thanks for the heads up.

File: torch/distributions/constraints.py [hunk 8]
```diff
@@ -275,6 +276,17 @@ def check(self, value):
...
 
 
+class _CorrCholesky(Constraint):
+    """
+    Constrain to lower-triangular square matrices with positive diagonals and each
+    row vector being of unit length.
+    """
+    def check(self, value):
+        row_norm = torch.linalg.norm(value, dim=-1)
+        unit_row_norm = ((row_norm <= 1.) & (row_norm >= (1 - 1e-6))).all(dim=-1)
```
Comment (fritzo) [inline torch/distributions/constraints.py:286]: I would expect this to fail for float32 and matrices of dim > 100. If you want to be safer you could try something like:
```py
tol = torch.finfo(value.dtype).eps * value.size(-1) * 10  # 10 is an arbitrary fudge factor
row_norm = torch.linalg.norm(value.detach(), dim=-1)
return (row_norm - 1).abs().le(tol).all(dim=-1)
```

**Review by fritzo — COMMENTED** [review]
Generally looks good to me, but I defer math checking to @fehiepsi :smile:
Thanks for refactoring the test files, they look great!

**Comment by neerajprad** [comment]
Thanks for the very helpful reviews, @fehiepsi, @fritzo. I'm noticing a few build failures and will address all your comments alongside that shortly.

File: torch/distributions/constraints.py (hunk 8 above)
Comment (neerajprad) [reply torch/distributions/constraints.py:286]: Sound advice! Thanks, will make the change.

Comment (neerajprad) [reply torch/distributions/constraints.py]: Done. One thing that I realized was that as the size of the matrix grows, the transformed matrix returned by `CorrCholeskyTransform` will have 0 diagonal entries and will violate the lower cholesky constraint. I am not entirely sure what to do about it since numerically this is bound to happen. cc @fehiepsi. 

```python
>>> y = CorrCholeskyTransform()(torch.randn(19900))  # 200 x 200 matrix
>>> y  # last rows all have 0 diagonal entries

tensor([[ 1.0000,  0.0000,  0.0000,  ...,  0.0000,  0.0000,  0.0000],
        [-0.6596,  0.7516,  0.0000,  ...,  0.0000,  0.0000,  0.0000],
        [ 0.3552,  0.4238,  0.8332,  ...,  0.0000,  0.0000,  0.0000],
        ...,
        [ 0.8225, -0.5501, -0.1322,  ...,  0.0000,  0.0000,  0.0000],
        [ 0.5494, -0.6964, -0.3553,  ...,  0.0000,  0.0000,  0.0000],
        [ 0.1776, -0.4688,  0.7778,  ..., -0.0000,  0.0000,  0.0000]])
>>> lower_cholesky.check(y)

tensor(False)

>>> y @ y.T  # return a random correlation matrix

tensor([[ 1.0000, -0.6596,  0.3552,  ...,  0.8225,  0.5494,  0.1776],
        [-0.6596,  1.0000,  0.0842,  ..., -0.9560, -0.8859, -0.4695],
        [ 0.3552,  0.0842,  1.0000,  ..., -0.0512, -0.3961,  0.5125],
        ...,
        [ 0.8225, -0.9560, -0.0512,  ...,  1.0000,  0.8932,  0.3093],
        [ 0.5494, -0.8859, -0.3961,  ...,  0.8932,  1.0000,  0.1874],
        [ 0.1776, -0.4695,  0.5125,  ...,  0.3093,  0.1874,  1.0000]])
```

Comment (neerajprad) [reply torch/distributions/constraints.py]: In any case, I think we should relax the constraint to >= 0 for diagonal entries to allow for correlation matrices that are positive semi definite.

Comment (fehiepsi) [reply torch/distributions/constraints.py]: I think we can replace
```
z1m_cumprod_sqrt = (1 - z).cumprod(-1).sqrt()
```
by
```
z1m_cumprod_sqrt = (1 - z).sqrt().cumprod(-1)
```
to improve precision but I am not sure if it will be stable enough. We might need to use float64 for a moderate dimension.

Comment (fehiepsi) [reply torch/distributions/constraints.py]: It seems to really help for matrices of up to 200 dimensions (diag entries can be as small as 1e-38). For large matrices, using float64 would be a better idea. :D

Comment (neerajprad) [reply torch/distributions/constraints.py]: Good idea, that should help. I'll make the change. If we have issues with larger correlation matrices, we can discuss work-arounds.

**Review by fehiepsi — APPROVED** [review]
LGTM, thanks @neerajprad!
